components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,515 @@
+  if (!uri.is_valid()) {
+    return false;
+  }
+  index += (value->first + 1);
+  uri_out = uri;
+  state_ = State::kVersion;
+  return true;
+}
+
+bool SIWEMessageParser::ParseVersion(std::string_view msg_view,
+                                     size_t& index,
+                                     uint32_t& version) {
+  auto value = ExtractValue(msg_view, kVersionToken);
+  if (!value) {
+    return false;
+  }
+  if (!base::StringToUint(value->second, &version)) {
+    return false;
+  }
+  // The only supported version should be 1
+  if (version != 1) {
+    return false;
+  }
+
+  index += (value->first + 1);
+  state_ = State::kChainId;
+  return true;
+}
+
+bool SIWEMessageParser::ParseChainId(std::string_view msg_view,
+                                     size_t& index,
+                                     uint64_t& chain_id) {
+  auto value = ExtractValue(msg_view, kChainIdToken);
+  if (!value) {
+    return false;
+  }
+  if (!base::StringToUint64(value->second, &chain_id)) {
+    return false;
+  }
+
+  index += (value->first + 1);
+  state_ = State::kNonce;
+  return true;
+}
+
+bool SIWEMessageParser::ParseNonce(std::string_view msg_view,
+                                   size_t& index,
+                                   std::string& nonce) {
+  auto value = ExtractValue(msg_view, kNonceToken);
+  if (!value) {
+    return false;
+  }
+  nonce = value->second;
+  if (nonce.size() < kMinNonceLength ||
+      !base::ranges::all_of(nonce, &base::IsAsciiAlphaNumeric<char>)) {
+    return false;
+  }
+
+  index += (value->first + 1);
+  state_ = State::kIssuedAt;
+  return true;
+}
+
+bool SIWEMessageParser::ParseIssuedAt(std::string_view msg_view,
+                                      size_t& index,
+                                      std::string& issued_at) {
+  // IssuedAt is the last required field so we have to check if it contains
+  // following optional fields
+  bool has_optional_fields = msg_view.rfind('\n') != std::string::npos;
+  auto value = ExtractValue(msg_view, kIssuedAtToken, has_optional_fields);
+  if (!value) {
+    return false;
+  }
+  issued_at = value->second;
+  index += value->first;
+  if (has_optional_fields) {
+    state_ = State::kOptionalFields;
+  } else {
+    state_ = State::kEnd;
+  }
+  return true;
+}
+
+bool SIWEMessageParser::ParseOptionalStringField(
+    std::string_view msg_view,
+    size_t& index,
+    const std::string& name,
+    absl::optional<std::string>& value_out) {
+  auto value = ExtractOptionalValue(msg_view, name);
+  if (!value) {
+    return false;
+  }
+  // Not presented
+  if (value->first == std::string::npos && value->second == msg_view) {
+    return true;
+  }
+  value_out = value->second;
+  index += value->first;
+  return true;
+}
+
+bool SIWEMessageParser::ParseOptionalResources(
+    std::string_view msg_view,
+    size_t& index,
+    absl::optional<std::vector<GURL>>& resources) {
+  if (msg_view.back() == '\n') {
+    return false;
+  }
```

Comment:
Not sure if it is guaranteed that msg_view is not empty here and accessing back() is correct

Reply:
`GetNextSegment` guarantee it won't be empty

Reply:
But it is better to check empty especially when we are moving toward https://github.com/brave/brave-core/pull/19443#discussion_r1322876395

Reply:
fixed in 438de51dfbf6ced8bb1bbffbcb7077529e4bc8b8